### PR TITLE
Type specialize Math.pow(int, int)

### PR DESCRIPTION
**Pull request by ThomsonTan**
Type specialize Math.pow for both magnitude and exponent, use powers of 2 to compute the exponentiation. Only integer integer multiplication involved. Both overflow and negative magnitude trigger bailout. Negative magnitude bails out because double is required to return and this scenario looks rare.

### CODE REVIEW

File: lib/Backend/BailOut.cpp [hunk 1]
```diff
@@ -2135,6 +2135,18 @@ void BailOutRecord::ScheduleFunctionCodeGen(Js::ScriptFunction * function, Js::S
                 }
                 break;
             }
+            case IR::BailOutOnPowIntIntOverflow:
```
Comment (rajatd) [inline lib/Backend/BailOut.cpp]: you should also add similar code in ScheduleLoopBodyCodeGen

File: lib/Backend/GlobOpt.cpp [hunk 2]
```diff
@@ -9591,13 +9594,29 @@ GlobOpt::TypeSpecializeInlineBuiltInBinary(IR::Instr **pInstr, Value *src1Val, V
...
 
+                this->ToInt32(instr, instr->GetSrc2(), this->currentBlock, src2Val, nullptr, lossy);
+
                 IR::Opnd* src1 = instr->GetSrc1();
-                this->ToFloat64(instr, src1, this->currentBlock, src1Val, nullptr, IR::BailOutPrimitiveButString);
+                int32 valueMin, valueMax;
+                if (src1Val->GetValueInfo()->IsLikelyInt() &&
+                    this->DoPowIntIntTypeSpec() &&
+                    src2Val->GetValueInfo()->GetIntValMinMax(&valueMin, &valueMax, this->DoAggressiveIntTypeSpec()) &&
+                    valueMax >= 0)
```
Comment (rajatd) [inline lib/Backend/GlobOpt.cpp]: shouldn't this be checking for `valueMin > 0`

**Comment by rajatd** [comment]
You should also update the method `GetBailOutKindName` to handle your new bailout kind

File: lib/Backend/GlobOpt.cpp (hunk 2 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOpt.cpp]: Good catch.

---
In reply to: [57935934](https://github.com/Microsoft/ChakraCore/pull/690#discussion_r57935934) [](ancestors = 57935934)

**Comment by rajatd** [comment]
:+1:

**Pull request merged**